airbyte-cdk/python/airbyte_cdk/sources/streams/core.py

```diff
@@ -123,6 +122,55 @@ def read_full_refresh(
                 cursor_field=cursor_field,
             )
 
+    def read_incremental(
+        self,
+        cursor_field: Optional[List[str]],
+        logger: logging.Logger,
+        slice_logger: SliceLogger,
+        stream_state: MutableMapping[str, Any],
+        state_manager,  # ignoring typing for ConnectorStateManager because of circular dependencies
+        per_stream_state_enabled: bool,
+        internal_config: InternalConfig,
+    ) -> Iterable[StreamData]:
+        slices = self.stream_slices(
+            cursor_field=cursor_field,
+            sync_mode=SyncMode.incremental,
+            stream_state=stream_state,
+        )
+        logger.debug(f"Processing stream slices for {self.name} (sync_mode: incremental)", extra={"stream_slices": slices})
+
+        has_slices = False
+        record_counter = 0
+        for _slice in slices:
+            has_slices = True
+            if slice_logger.should_log_slice_message(logger):
+                yield slice_logger.create_slice_log_message(_slice)
+            records = self.read_records(
+                sync_mode=SyncMode.incremental,
+                stream_slice=_slice,
+                stream_state=stream_state,
+                cursor_field=cursor_field or None,
+            )
+            for record_data_or_message in records:
+                yield record_data_or_message
+                if isinstance(record_data_or_message, Mapping) or (hasattr(record_data_or_message, "type") and record_data_or_message.type == MessageType.RECORD):
+                    record_data = record_data_or_message if isinstance(record_data_or_message, Mapping) else record_data_or_message.record
+                    stream_state = self.get_updated_state(stream_state, record_data)
+                    checkpoint_interval = self.state_checkpoint_interval
+                    record_counter += 1
+                    if checkpoint_interval and record_counter % checkpoint_interval == 0:
+                        yield self._checkpoint_state(stream_state, state_manager, per_stream_state_enabled)
+
+                    if internal_config.is_limit_reached(record_counter):
```

Comment:
I don't like the discrepancy where for full_refresh, this logic is in `abstract_source` and for incremental, it's here. The reason I had to more it here is that if we hit the record limit, we have to emit a state message hence, the logic can't be separated from the state management

Reply:
Based on [your comment](https://github.com/airbytehq/airbyte/pull/31466/files#r1370827267) we're unsure about the future of this thing. I wouldn't waste time moving things are if this is something we're not sure we'll maintain. I'll take a note on my side to investigate the internal_config limit and create an issue to clean it up if necessary

Reply:
I've checked for the use of `internal_config` and based on the secrets I've pulled locally, `airbyte-integrations/connectors/source-salesforce/secrets/config_bulk.json` has `_page_size` as a config key and hence uses the internal config which is used for `HttpStream`. However, I don't see this config being reference in CATs... I'll ask [more questions](https://airbytehq-team.slack.com/archives/C02U9R3AF37/p1698679523006099)

Reply:
Confirmed that this is not used. We will prioritize https://github.com/airbytehq/airbyte/issues/30088